design/proof-erasure-coding.md

```diff
@@ -0,0 +1,419 @@
+matter how big the shards are.
+
+The amount of original data shards is typically called K, the amount of parity
+shards M, and the total amount of shards N.
+
+Adversarial erasure
+-------------------
+
+The disadvantage of interleaving is that it weakens the protection against
+adversarial erasure that Reed-Solomon provides.
+
+An adversary can now strategically remove only the first element from more than
+half of the shards, and the dataset will be damaged beyond repair. For example,
+with a dataset of 1TB erasure coded into 256 data and parity shards, an
+adversary could strategically remove 129 bytes, and the data can no longer be
+fully recovered.
+
+Implications for storage proofs
+-------------------------------
+
+This means that when we check for missing data, we should perform our checks on
+entire shards to protect against adversarial erasure. In the case of our Merkle
+storage proofs, this means that we need to hash the entire shard, and then check
+that hash with a Merkle proof. Effectively the block size for Merkle proofs
+should equal the shard size of the erasure coding interleaving. This is rather
+unfortunate, because hashing large amounts of data is rather expensive to
+perform in a SNARK, which is what we use to compress proofs in size.
+
+A large amount of input data in a SNARK leads to a larger circuit, and to more
+iterations of the hashing algorithm, which also leads to a larger circuit. A
+larger circuit means longer computation and higher memory consumption.
+
+Ideally, we'd like to have small blocks to keep Merkle proofs inside SNARKs
+relatively performant, but we are limited by the maximum amount of shards that a
+particular Reed-Solomon algorithm supports. For instance, the [leopard][1]
+library can create at most 65536 shards, because it uses a Galois field of 2^16.
+Should we use this to encode a 1TB file, we'd end up with shards of 16MB, far
+too large to be practical in a SNARK.
+
+Design space
+------------
+
+This limits the choices that we can make. The limiting factors seem to be:
+
+- Maximum number of shards, determined by the field size of the erasure coding
+  algorithm
+- Number of blocks per proof, which determines how likely we are to detect
+  missing blocks
+- Capacity of the SNARK algorithm; how many bytes can we hash in a reasonable
+  time inside the SNARK
+
+From these limiting factors we can derive:
+
+- Block size; equals shard size
+- Maximum slot size; the maximum amount of data that can be verified with a
+  proof
+- Erasure coding memory requirements
+
+For example, when we use the leopard library, with a Galois field of 2^16, and
+require 80 blocks to be sampled per proof, and we can implement a SNARK that can
+hash 80*64K bytes, then we have:
+
+- Block size: 64KB
+- Maximum slot size: 4GB (2^16 * 64KB)
+- Erasure coding memory: > 128KB (2^16 * 16 bits)
+
+Which has the disadvantage of having a rather low maximum slot size of 4GB. When
```

Comment:
As I understood it, the 1TB slot size comes from tokenomics. Providing a proof on-chain costs money, and it's not economically viable to do this for a large number of 4GB slots. It is however viable for 1TB slots.

Reply:
I worried more about proof creation costs than proof verification costs when came up with the 1 TB number. On-chain verification is expensive, true, but in theory we will be able to aggregate proofs (otherwise the whole system is unfeasible). However, if a node has say 4 TB of data (not a very big storage node), then it has to produce (4TB / <slot_ size>) proofs. My back-of-the-envelope calculation for a single proof with the current proposal gave approx 1 minute of proof time (single core); this limits the number of proofs a node can produce in a day. Also if the node spends too much time computing proofs then its energy cost will be presumably too high.